contracts/BatchExchangeViewer.sol

```diff
@@ -0,0 +1,121 @@
+     *  @return encoded bytes representing orders
+     */
+    function getFinalizedOrderBook() public view returns (bytes memory) {
+        (bytes memory elements, , ) = getFinalizedOrderBookPaginated(address(0), 0, uint16(-1));
+        return elements;
+    }
+
+    /** @dev Queries a page of the orderbook for the auction that is currently being solved
+     *  @param previousPageUser address taken from nextPageUser return value from last page (address(0) for first page)
+     *  @param previousPageUserOffset offset taken nextPageUserOffset return value from last page (0 for first page)
+     *  @param pageSize count of elements to be returned per page (same value is used for subqueries on the exchange)
+     *  @return encoded bytes representing orders and page information for next page
+     */
+    function getFinalizedOrderBookPaginated(address previousPageUser, uint16 previousPageUserOffset, uint16 pageSize)
+        public
+        view
+        returns (bytes memory elements, address nextPageUser, uint16 nextPageUserOffset)
+    {
+        uint32 batch = batchExchange.getCurrentBatchId();
+        return getEncodedOrdersPaginated(batch - 1, batch - 1, previousPageUser, previousPageUserOffset, pageSize);
+    }
+
+    /** @dev Queries a page in the list of all orders
+     *  @param maxValidFrom all returned orders will have a validFrom <= this value (they were placed at or before that batch)
+     *  @param minValidUntil all returned orders will have a validUntil >= this value (validity ends at or after that batch)
+     *  @param previousPageUser address taken from nextPageUser return value from last page (address(0) for first page)
+     *  @param previousPageUserOffset offset taken nextPageUserOffset return value from last page (0 for first page)
+     *  @param pageSize count of elements to be returned per page (same value is used for subqueries on the exchange)
+     *  @return encoded bytes representing orders and page information for next page
+     */
+    function getEncodedOrdersPaginated(
+        uint32 maxValidFrom,
+        uint32 minValidUntil,
+        address previousPageUser,
+        uint16 previousPageUserOffset,
+        uint16 pageSize
+    ) public view returns (bytes memory elements, address nextPageUser, uint16 nextPageUserOffset) {
+        nextPageUser = previousPageUser;
+        nextPageUserOffset = previousPageUserOffset;
+        bool hasNextPage = true;
+        while (hasNextPage) {
+            bytes memory unfiltered = batchExchange.getEncodedUsersPaginated(nextPageUser, nextPageUserOffset, pageSize);
+            hasNextPage = unfiltered.length / AUCTION_ELEMENT_WIDTH == pageSize;
+            for (uint16 index = 0; index < unfiltered.length / AUCTION_ELEMENT_WIDTH; index++) {
+                bytes memory element = unfiltered.slice(index * AUCTION_ELEMENT_WIDTH, AUCTION_ELEMENT_WIDTH);
+                if (maxValidFrom >= getValidFrom(element) && minValidUntil <= getValidUntil(element)) {
```

Comment:
If gas is a concern, it looks like it might make sense to join `getValidFrom` and `getValidUntil` into one single call. I there's no notable gas difference, I like better your approach